### ISSUE TITLE
Update `no_default_cases` to cover exhaustive switches

### ISSUE THREAD

**Issue by nshahan** [dart-lang/linter#4762]
Exhaustive switches over sealed class hierarchies were added in Dart 3. 

In some cases developers want a compile time error in the event new classes are added to that hierarchy and are not explicitly handled by a switch case. This means we need some way to prevent writing default. It seems like `no_default_cases` is a prime candidate for updating to handle this new behavior.

**To Reproduce**
```dart
sealed class A {}
class B implements A {}
class C implements A {}

void fn(A a) {
  switch (a) {
    case B():
      print('B');
    default:
      print('default');
  }
}
```

**Expected behavior**
I would like to see a lint on the `default` case explaining that each case of the sealed class hierarchy should be written explicitly.

**Additional context**
Somewhat related issues:
https://github.com/dart-lang/linter/issues/4735
https://github.com/dart-lang/linter/issues/3967

**Comment by jakemac53** [comment]
It definitely seems appropriate for this lint to also apply to sealed classes, given exhaustiveness is their primary benefit and purpose for existing :)

**Comment by goderbauer** [comment]
I would have totally expected to see the lint fire on the example given.

**Comment by munificent** [comment]
Extending `no_default_cases` to sealed classes seems reasonable to me. Though it's worth thinking about how this interacts with more complex uses of pattern matching and exhaustiveness. Should this garner a lint error:

```dart
sealed class A {}
class B implements A {}
class C implements A {}

void fn(A a) {
  switch (a) {
    case B():
      print('B');
    case _: // <-- Wildcard pattern.
      print('default');
  }
}
```

What about:

```dart
void fn(A a1, A a2) {
  switch ((a1, a2)) {
    case (B(), B()):
      print('Two Bs');
    case (_, _):
      print('default');
  }
}
```

Or:

```dart
sealed class A {}
class B implements A {}
class C implements A {}

void fn(A a) {
  switch (a) {
    case B():
      print('B');
    case A(): // <-- Always matches even if new subtypes are added.
      print('default');
  }
}
```

Personally, I lean towards letting users decide how they want to ensure their switches are exhaustive. If they want to know when new cases are added, they can list all the subtypes explicitly. If they don't, they can match the whole supertype.
